In [1]:
import numpy as np
import pandas as pd

In [4]:
# numpy version check
print(np.__version__)
# Pandas version
print(pd.__version__)

1.16.5
0.25.1


In [2]:
# Read the data
data = pd.read_csv('sample.csv')

In [3]:
# Checking output type
print(type(data))

<class 'pandas.core.frame.DataFrame'>


In [5]:
# To check the shape
print((data.shape))
# to get the number of rows or instances or records or examples
print(data.shape[0])
# to get the no of columns or features or attributes
print(data.shape[1])

(756, 18)
756
18


In [ ]:
# To get the columns names
print(data.columns)
# to get the no of rows or instancces or records or examples
print(data.shape[0])
# to get the columns names
print(data.columns)
# to get the row names
print(data.index)

In [6]:
# Data type of each columns
print(data.dtypes)
print('......'*15)
# what is the output type for datatypes
print(type(data.dtypes))
print('.......'*15)
# Access the first column dtype
print(data.dtypes[0])
print('........'*15)
# Access the last column dtype
print(data.dtypes[-1])

id                  int64
season              int64
city               object
date               object
team1              object
team2              object
toss_winner        object
toss_decision      object
result             object
dl_applied          int64
winner             object
win_by_runs         int64
win_by_wickets      int64
player_of_match    object
venue              object
umpire1            object
umpire2            object
umpire3            object
dtype: object
..........................................................................................
<class 'pandas.core.series.Series'>
.........................................................................................................
int64
........................................................................................................................
object


In [79]:
# Column count for each datatype
# How many columns are there with int64
# How many cloumns are there with float64
# How many columns are there with object 
data.dtypes.value_counts()
# Note We are using series method value_counts

object    13
int64      5
dtype: int64

### Observe the data

In [8]:
#display the top5 rows
data.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [7]:
#display the top 10 rows
data.head(10)

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   
5   6    2017  Hyderabad  2017-04-09                Gujarat Lions   
6   7    2017     Mumbai  2017-04-09        Kolkata Knight Riders   
7   8    2017     Indore  2017-04-10  Royal Challengers Bangalore   
8   9    2017       Pune  2017-04-11             Delhi Daredevils   
9  10    2017     Mumbai  2017-04-12          Sunrisers Hyderabad   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   
5          Sunrisers Hyderabad          Sunrisers Hyderabad         field   
6               Mumbai Indians               Mumbai Indians         field   
7              Kings XI Punjab  Royal Challengers Bangalore           bat   
8       Rising Pune Supergiant       Rising Pune Supergiant         field   
9               Mumbai Indians               Mumbai Indians         field   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   
5  normal           0          Sunrisers Hyderabad            0   
6  normal           0               Mumbai Indians            0   
7  normal           0              Kings XI Punjab            0   
8  normal           0             Delhi Daredevils           97   
9  normal           0               Mumbai Indians            0   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   
5               9     Rashid Khan  Rajiv Gandhi International Stadium, Uppal   
6               4          N Rana                           Wankhede Stadium   
7               8        AR Patel                     Holkar Cricket Stadium   
8               0       SV Samson    Maharashtra Cricket Association Stadium   
9               4       JJ Bumrah                           Wankhede Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN  
5      A Deshmukh       NJ Llong     NaN  
6     Nitin Menon      CK Nandan     NaN  
7    AK Chaudhary  C Shamshuddin     NaN  
8     AY Dandekar         S Ravi     NaN  
9     Nitin Menon      CK Nandan     NaN

In [9]:
# select 5 rows randomly
data.sample(5)

id  season     city        date                        team1  \
341    342    2012    Delhi  2012-04-27             Delhi Daredevils   
185    186    2010  Cuttack  2010-03-19              Deccan Chargers   
555    556    2015   Mumbai  2015-05-05             Delhi Daredevils   
303    304    2011   Mumbai  2011-05-24  Royal Challengers Bangalore   
735  11331    2019   Jaipur    22/04/19             Rajasthan Royals   

                   team2          toss_winner toss_decision  result  \
341       Mumbai Indians       Mumbai Indians         field  normal   
185      Kings XI Punjab      Kings XI Punjab         field  normal   
555       Mumbai Indians     Delhi Daredevils           bat  normal   
303  Chennai Super Kings  Chennai Super Kings         field  normal   
735       Delhi Capitals       Delhi Capitals         field  normal   

     dl_applied               winner  win_by_runs  win_by_wickets  \
341           0     Delhi Daredevils           37               0   
185           0      Deccan Chargers            6               0   
555           0       Mumbai Indians            0               5   
303           0  Chennai Super Kings            0               6   
735           0       Delhi Capitals            0               6   

     player_of_match                   venue          umpire1        umpire2  \
341         V Sehwag        Feroz Shah Kotla        Aleem Dar   BNJ Oxenford   
185        A Symonds        Barabati Stadium        BF Bowden      M Erasmus   
555  Harbhajan Singh        Wankhede Stadium  HDPK Dharmasena    CB Gaffaney   
303         SK Raina        Wankhede Stadium        Asad Rauf     SJA Taufel   
735          RR Pant  Sawai Mansingh Stadium           S Ravi  Nanda Kishore   

            umpire3  
341             NaN  
185             NaN  
555             NaN  
303             NaN  
735  Yeshwant Barde

In [10]:
# dispaly the last 3 rows
data.tail(3)

id  season           city      date                team1  \
753  11413    2019  Visakhapatnam  08/05/19  Sunrisers Hyderabad   
754  11414    2019  Visakhapatnam  10/05/19       Delhi Capitals   
755  11415    2019      Hyderabad  12/05/19       Mumbai Indians   

                   team2          toss_winner toss_decision  result  \
753       Delhi Capitals       Delhi Capitals         field  normal   
754  Chennai Super Kings  Chennai Super Kings         field  normal   
755  Chennai Super Kings       Mumbai Indians           bat  normal   

     dl_applied               winner  win_by_runs  win_by_wickets  \
753           0       Delhi Capitals            0               2   
754           0  Chennai Super Kings            0               6   
755           0       Mumbai Indians            1               0   

    player_of_match                               venue        umpire1  \
753         RR Pant                    ACA-VDCA Stadium            NaN   
754    F du Plessis                    ACA-VDCA Stadium  Sundaram Ravi   
755       JJ Bumrah  Rajiv Gandhi Intl. Cricket Stadium    Nitin Menon   

            umpire2                  umpire3  
753             NaN                      NaN  
754  Bruce Oxenford  Chettithody Shamshuddin  
755       Ian Gould              Nigel Llong

In [11]:
#select a random row
data.sample()

id  season    city        date             team1                  team2  \
77  78    2008  Jaipur  2008-05-01  Rajasthan Royals  Kolkata Knight Riders   

         toss_winner toss_decision  result  dl_applied            winner  \
77  Rajasthan Royals           bat  normal           0  Rajasthan Royals   

    win_by_runs  win_by_wickets player_of_match                   venue  \
77           45               0     SA Asnodkar  Sawai Mansingh Stadium   

        umpire1         umpire2 umpire3  
77  RE Koertzen  GA Pratapkumar     NaN

- data.head()
- data.tail()
- data.sample()

the output type of all the above methods are dataframe

In [14]:
data.head(1)

id  season       city        date                team1  \
0   1    2017  Hyderabad  2017-04-05  Sunrisers Hyderabad   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   

   result  dl_applied               winner  win_by_runs  win_by_wickets  \
0  normal           0  Sunrisers Hyderabad           35               0   

  player_of_match                                      venue      umpire1  \
0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal  AY Dandekar   

    umpire2 umpire3  
0  NJ Llong     NaN

In [15]:
data.tail(1)

id  season       city      date           team1                team2  \
755  11415    2019  Hyderabad  12/05/19  Mumbai Indians  Chennai Super Kings   

        toss_winner toss_decision  result  dl_applied          winner  \
755  Mumbai Indians           bat  normal           0  Mumbai Indians   

     win_by_runs  win_by_wickets player_of_match  \
755            1               0       JJ Bumrah   

                                  venue      umpire1    umpire2      umpire3  
755  Rajiv Gandhi Intl. Cricket Stadium  Nitin Menon  Ian Gould  Nigel Llong

- data.head(1)
- data.sample(1)

the putput type of all the above methods are also dataframe which consists a single row

### Info about the dataframe

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
id                 756 non-null int64
season             756 non-null int64
city               749 non-null object
date               756 non-null object
team1              756 non-null object
team2              756 non-null object
toss_winner        756 non-null object
toss_decision      756 non-null object
result             756 non-null object
dl_applied         756 non-null int64
winner             752 non-null object
win_by_runs        756 non-null int64
win_by_wickets     756 non-null int64
player_of_match    752 non-null object
venue              756 non-null object
umpire1            754 non-null object
umpire2            754 non-null object
umpire3            119 non-null object
dtypes: int64(5), object(13)
memory usage: 106.4+ KB


In [17]:
type(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
id                 756 non-null int64
season             756 non-null int64
city               749 non-null object
date               756 non-null object
team1              756 non-null object
team2              756 non-null object
toss_winner        756 non-null object
toss_decision      756 non-null object
result             756 non-null object
dl_applied         756 non-null int64
winner             752 non-null object
win_by_runs        756 non-null int64
win_by_wickets     756 non-null int64
player_of_match    752 non-null object
venue              756 non-null object
umpire1            754 non-null object
umpire2            754 non-null object
umpire3            119 non-null object
dtypes: int64(5), object(13)
memory usage: 106.4+ KB


NoneType

### Accesing a single column

In [18]:
data.columns

Index(['id', 'season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [19]:
# To access the column id
data.id

0          1
1          2
2          3
3          4
4          5
       ...  
751    11347
752    11412
753    11413
754    11414
755    11415
Name: id, Length: 756, dtype: int64

In [21]:
# What is the output type
print(type(data.id))

<class 'pandas.core.series.Series'>


In [20]:
# To access the column id
data['id']

0          1
1          2
2          3
3          4
4          5
       ...  
751    11347
752    11412
753    11413
754    11414
755    11415
Name: id, Length: 756, dtype: int64

In [22]:
# What is the putput type
print(type(data['id']))

<class 'pandas.core.series.Series'>


### Specific column Datatype

In [24]:
data.season

0      2017
1      2017
2      2017
3      2017
4      2017
       ... 
751    2019
752    2019
753    2019
754    2019
755    2019
Name: season, Length: 756, dtype: int64

In [27]:
# specific column datatype
data.season.dtype

dtype('int64')

### difference bt dtype and dtypes
- dataframe object doesnt have dtype

- Numerical columns  - int or float data
    - mean , meadian , sum.quartiles , min, max
- Categorical columns -  object type or string type data
- Bollean data - True or False

In [28]:
# mean value of a column
data.season.mean()

2013.4444444444443

In [29]:
# sum value of a column
data.season.sum()/data.season.shape[0]

2013.4444444444443

In [30]:
# min value of a column
print(data.season.min())
# max value of a column
print(data.season.max())

2008
2019


In [31]:
# mode means the value which is repeating that value wil be displayed
data.season.mode()

0    2013
dtype: int64

In [32]:
data.season.value_counts()

2013    76
2012    74
2011    73
2019    60
2018    60
2016    60
2014    60
2010    60
2017    59
2015    59
2008    58
2009    57
Name: season, dtype: int64

In [33]:
data.season.mode()[0]

2013

In [34]:
data.season.median()

2013.0

In [35]:
# Range of values in a column
(data.season.min(),data.season.max())

(2008, 2019)

In [39]:
data.season.count()

756

In [36]:
data.columns

Index(['id', 'season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [37]:
data.city.dtype

dtype('O')

In [38]:
# How many no nul values are there
data.city.count()

749

In [40]:
# For each unique value what is the count
data.city.value_counts()
# output type is series
# In that series the lables are unique values
# the value for that label is how many times it occured in that column
# the final dtype refers to the value type of the series count is integer

Mumbai            101
Kolkata            77
Delhi              74
Bangalore          66
Hyderabad          64
Chennai            57
Jaipur             47
Chandigarh         46
Pune               38
Durban             15
Bengaluru          14
Visakhapatnam      13
Centurion          12
Ahmedabad          12
Rajkot             10
Mohali             10
Dharamsala          9
Indore              9
Johannesburg        8
Port Elizabeth      7
Cuttack             7
Abu Dhabi           7
Cape Town           7
Ranchi              7
Sharjah             6
Raipur              6
Kochi               5
Kanpur              4
Kimberley           3
Nagpur              3
East London         3
Bloemfontein        2
Name: city, dtype: int64

In [43]:
#  To check  how many Unique values are there
print(data.city.nunique())
print(data.city.value_counts().shape[0])

32
32


In [45]:
# Wht are the unique values 
print(data.city.unique())
print(data.city.value_counts().index)

['Hyderabad' 'Pune' 'Rajkot' 'Indore' 'Bangalore' 'Mumbai' 'Kolkata'
 'Delhi' 'Chandigarh' 'Kanpur' 'Jaipur' 'Chennai' 'Cape Town'
 'Port Elizabeth' 'Durban' 'Centurion' 'East London' 'Johannesburg'
 'Kimberley' 'Bloemfontein' 'Ahmedabad' 'Cuttack' 'Nagpur' 'Dharamsala'
 'Kochi' 'Visakhapatnam' 'Raipur' 'Ranchi' 'Abu Dhabi' 'Sharjah' nan
 'Mohali' 'Bengaluru']
Index(['Mumbai', 'Kolkata', 'Delhi', 'Bangalore', 'Hyderabad', 'Chennai',
       'Jaipur', 'Chandigarh', 'Pune', 'Durban', 'Bengaluru', 'Visakhapatnam',
       'Centurion', 'Ahmedabad', 'Rajkot', 'Mohali', 'Dharamsala', 'Indore',
       'Johannesburg', 'Port Elizabeth', 'Cuttack', 'Abu Dhabi', 'Cape Town',
       'Ranchi', 'Sharjah', 'Raipur', 'Kochi', 'Kanpur', 'Kimberley', 'Nagpur',
       'East London', 'Bloemfontein'],
      dtype='object')


In [46]:
# To get the description for the numerical columns
data.describe()

id       season  dl_applied  win_by_runs  win_by_wickets
count    756.000000   756.000000  756.000000   756.000000      756.000000
mean    1792.178571  2013.444444    0.025132    13.283069        3.350529
std     3464.478148     3.366895    0.156630    23.471144        3.387963
min        1.000000  2008.000000    0.000000     0.000000        0.000000
25%      189.750000  2011.000000    0.000000     0.000000        0.000000
50%      378.500000  2013.000000    0.000000     0.000000        4.000000
75%      567.250000  2016.000000    0.000000    19.000000        6.000000
max    11415.000000  2019.000000    1.000000   146.000000       10.000000

In [49]:
# To get the list of numerical columns
data.describe().columns

Index(['id', 'season', 'dl_applied', 'win_by_runs', 'win_by_wickets'], dtype='object')

In [50]:
# To get the list of categorical columns
data.describe(include=object)

city        date           team1                        team2  \
count      749         756             756                          756   
unique      32         546              15                           15   
top     Mumbai  2012-04-22  Mumbai Indians  Royal Challengers Bangalore   
freq       101           2             101                           95   

           toss_winner toss_decision  result          winner player_of_match  \
count              756           756     756             752             752   
unique              15             2       3              15             226   
top     Mumbai Indians         field  normal  Mumbai Indians        CH Gayle   
freq                98           463     743             109              21   

               venue          umpire1        umpire2      umpire3  
count            756              754            754          119  
unique            41               61             65           25  
top     Eden Gardens  HDPK Dharmasena  C Shamshuddin  Nitin Menon  
freq              77               73             57           10

In [51]:
data.describe(include=np.int64)

id       season  dl_applied  win_by_runs  win_by_wickets
count    756.000000   756.000000  756.000000   756.000000      756.000000
mean    1792.178571  2013.444444    0.025132    13.283069        3.350529
std     3464.478148     3.366895    0.156630    23.471144        3.387963
min        1.000000  2008.000000    0.000000     0.000000        0.000000
25%      189.750000  2011.000000    0.000000     0.000000        0.000000
50%      378.500000  2013.000000    0.000000     0.000000        4.000000
75%      567.250000  2016.000000    0.000000    19.000000        6.000000
max    11415.000000  2019.000000    1.000000   146.000000       10.000000

In [55]:
data.describe(include='all')

id       season    city        date           team1  \
count     756.000000   756.000000     749         756             756   
unique           NaN          NaN      32         546              15   
top              NaN          NaN  Mumbai  2012-04-22  Mumbai Indians   
freq             NaN          NaN     101           2             101   
mean     1792.178571  2013.444444     NaN         NaN             NaN   
std      3464.478148     3.366895     NaN         NaN             NaN   
min         1.000000  2008.000000     NaN         NaN             NaN   
25%       189.750000  2011.000000     NaN         NaN             NaN   
50%       378.500000  2013.000000     NaN         NaN             NaN   
75%       567.250000  2016.000000     NaN         NaN             NaN   
max     11415.000000  2019.000000     NaN         NaN             NaN   

                              team2     toss_winner toss_decision  result  \
count                           756             756           756     756   
unique                           15              15             2       3   
top     Royal Challengers Bangalore  Mumbai Indians         field  normal   
freq                             95              98           463     743   
mean                            NaN             NaN           NaN     NaN   
std                             NaN             NaN           NaN     NaN   
min                             NaN             NaN           NaN     NaN   
25%                             NaN             NaN           NaN     NaN   
50%                             NaN             NaN           NaN     NaN   
75%                             NaN             NaN           NaN     NaN   
max                             NaN             NaN           NaN     NaN   

        dl_applied          winner  win_by_runs  win_by_wickets  \
count   756.000000             752   756.000000      756.000000   
unique         NaN              15          NaN             NaN   
top            NaN  Mumbai Indians          NaN             NaN   
freq           NaN             109          NaN             NaN   
mean      0.025132             NaN    13.283069        3.350529   
std       0.156630             NaN    23.471144        3.387963   
min       0.000000             NaN     0.000000        0.000000   
25%       0.000000             NaN     0.000000        0.000000   
50%       0.000000             NaN     0.000000        4.000000   
75%       0.000000             NaN    19.000000        6.000000   
max       1.000000             NaN   146.000000       10.000000   

       player_of_match         venue          umpire1        umpire2  \
count              752           756              754            754   
unique             226            41               61             65   
top           CH Gayle  Eden Gardens  HDPK Dharmasena  C Shamshuddin   
freq                21            77               73             57   
mean               NaN           NaN              NaN            NaN   
std                NaN           NaN              NaN            NaN   
min                NaN           NaN              NaN            NaN   
25%                NaN           NaN              NaN            NaN   
50%                NaN           NaN              NaN            NaN   
75%                NaN           NaN              NaN            NaN   
max                NaN           NaN              NaN            NaN   

            umpire3  
count           119  
unique           25  
top     Nitin Menon  
freq             10  
mean            NaN  
std             NaN  
min             NaN  
25%             NaN  
50%             NaN  
75%             NaN  
max             NaN

#### Slicing, Subsetting, Boolean Indexing
- Positioned based indexing
    - rows 0 to 756
-  label based indexing
    - rows 0 to 756
    - columns - ['id', 'season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3']

In [56]:
data.iloc[1,:]# row with index 1 and all athe cloumns

id                                                       2
season                                                2017
city                                                  Pune
date                                            2017-04-06
team1                                       Mumbai Indians
team2                               Rising Pune Supergiant
toss_winner                         Rising Pune Supergiant
toss_decision                                        field
result                                              normal
dl_applied                                               0
winner                              Rising Pune Supergiant
win_by_runs                                              0
win_by_wickets                                           7
player_of_match                                  SPD Smith
venue              Maharashtra Cricket Association Stadium
umpire1                                     A Nand Kishore
umpire2                                             S Ra

In [60]:
data.iloc[1,1] # row with index 1 and cloumn with index1

2017

In [61]:
data.loc[1,'season']

2017

In [63]:
data.iloc[1,1:3] # row with index 1 and columns with index1,2
# iloc stop index is exclusive

season    2017
city      Pune
Name: 1, dtype: object

In [64]:
data.loc[1,'season':'city']
# loc -  the stop label also inclusive

season    2017
city      Pune
Name: 1, dtype: object

In [65]:
data.iloc[1:4,1:3]
# rows iwth indices 1,2,3
# cloumns with indices 1,2

season    city
1    2017    Pune
2    2017  Rajkot
3    2017  Indore

In [66]:
data.loc[1:3,'season':'city']

season    city
1    2017    Pune
2    2017  Rajkot
3    2017  Indore

In [67]:
data.id.nunique()

756

In [69]:
data.set_index('id')

season           city        date                        team1  \
id                                                                      
1        2017      Hyderabad  2017-04-05          Sunrisers Hyderabad   
2        2017           Pune  2017-04-06               Mumbai Indians   
3        2017         Rajkot  2017-04-07                Gujarat Lions   
4        2017         Indore  2017-04-08       Rising Pune Supergiant   
5        2017      Bangalore  2017-04-08  Royal Challengers Bangalore   
...       ...            ...         ...                          ...   
11347    2019         Mumbai    05/05/19        Kolkata Knight Riders   
11412    2019        Chennai    07/05/19          Chennai Super Kings   
11413    2019  Visakhapatnam    08/05/19          Sunrisers Hyderabad   
11414    2019  Visakhapatnam    10/05/19               Delhi Capitals   
11415    2019      Hyderabad    12/05/19               Mumbai Indians   

                             team2                  toss_winner toss_decision  \
id                                                                              
1      Royal Challengers Bangalore  Royal Challengers Bangalore         field   
2           Rising Pune Supergiant       Rising Pune Supergiant         field   
3            Kolkata Knight Riders        Kolkata Knight Riders         field   
4                  Kings XI Punjab              Kings XI Punjab         field   
5                 Delhi Daredevils  Royal Challengers Bangalore           bat   
...                            ...                          ...           ...   
11347               Mumbai Indians               Mumbai Indians         field   
11412               Mumbai Indians          Chennai Super Kings           bat   
11413               Delhi Capitals               Delhi Capitals         field   
11414          Chennai Super Kings          Chennai Super Kings         field   
11415          Chennai Super Kings               Mumbai Indians           bat   

       result  dl_applied                       winner  win_by_runs  \
id                                                                    
1      normal           0          Sunrisers Hyderabad           35   
2      normal           0       Rising Pune Supergiant            0   
3      normal           0        Kolkata Knight Riders            0   
4      normal           0              Kings XI Punjab            0   
5      normal           0  Royal Challengers Bangalore           15   
...       ...         ...                          ...          ...   
11347  normal           0               Mumbai Indians            0   
11412  normal           0               Mumbai Indians            0   
11413  normal           0               Delhi Capitals            0   
11414  normal           0          Chennai Super Kings            0   
11415  normal           0               Mumbai Indians            1   

       win_by_wickets player_of_match  \
id                                      
1                   0    Yuvraj Singh   
2                   7       SPD Smith   
3                  10         CA Lynn   
4                   6      GJ Maxwell   
5                   0       KM Jadhav   
...               ...             ...   
11347               9       HH Pandya   
11412               6        AS Yadav   
11413               2         RR Pant   
11414               6    F du Plessis   
11415               0       JJ Bumrah   

                                           venue         umpire1  \
id                                                                 
1      Rajiv Gandhi International Stadium, Uppal     AY Dandekar   
2        Maharashtra Cricket Association Stadium  A Nand Kishore   
3         Saurashtra Cricket Association Stadium     Nitin Menon   
4                         Holkar Cricket Stadium    AK Chaudhary   
5                          M Chinnaswamy Stadium             NaN   
...                                          ...             ...   
1

In [70]:
data

id  season           city        date                        team1  \
0        1    2017      Hyderabad  2017-04-05          Sunrisers Hyderabad   
1        2    2017           Pune  2017-04-06               Mumbai Indians   
2        3    2017         Rajkot  2017-04-07                Gujarat Lions   
3        4    2017         Indore  2017-04-08       Rising Pune Supergiant   
4        5    2017      Bangalore  2017-04-08  Royal Challengers Bangalore   
..     ...     ...            ...         ...                          ...   
751  11347    2019         Mumbai    05/05/19        Kolkata Knight Riders   
752  11412    2019        Chennai    07/05/19          Chennai Super Kings   
753  11413    2019  Visakhapatnam    08/05/19          Sunrisers Hyderabad   
754  11414    2019  Visakhapatnam    10/05/19               Delhi Capitals   
755  11415    2019      Hyderabad    12/05/19               Mumbai Indians   

                           team2                  toss_winner toss_decision  \
0    Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1         Rising Pune Supergiant       Rising Pune Supergiant         field   
2          Kolkata Knight Riders        Kolkata Knight Riders         field   
3                Kings XI Punjab              Kings XI Punjab         field   
4               Delhi Daredevils  Royal Challengers Bangalore           bat   
..                           ...                          ...           ...   
751               Mumbai Indians               Mumbai Indians         field   
752               Mumbai Indians          Chennai Super Kings           bat   
753               Delhi Capitals               Delhi Capitals         field   
754          Chennai Super Kings          Chennai Super Kings         field   
755          Chennai Super Kings               Mumbai Indians           bat   

     result  dl_applied                       winner  win_by_runs  \
0    normal           0          Sunrisers Hyderabad           35   
1    normal           0       Rising Pune Supergiant            0   
2    normal           0        Kolkata Knight Riders            0   
3    normal           0              Kings XI Punjab            0   
4    normal           0  Royal Challengers Bangalore           15   
..      ...         ...                          ...          ...   
751  normal           0               Mumbai Indians            0   
752  normal           0               Mumbai Indians            0   
753  normal           0               Delhi Capitals            0   
754  normal           0          Chennai Super Kings            0   
755  normal           0               Mumbai Indians            1   

     win_by_wickets player_of_match  \
0                 0    Yuvraj Singh   
1                 7       SPD Smith   
2                10         CA Lynn   
3                 6      GJ Maxwell   
4                 0       KM Jadhav   
..              ...             ...   
751               9       HH Pandya   
752               6        AS Yadav   
753               2         RR Pant   
754               6    F du Plessis   
755               0       JJ Bumrah   

                                         venue         umpire1  \
0    Rajiv Gandhi International Stadium, Uppal     AY Dandekar   
1      Maharashtra Cricket Association Stadium  A Nand Kishore   
2       Saurashtra Cricket Association Stadium     Nitin Menon   
3                       Holkar Cricket Stadium    AK Chaudhary   
4                        M Chinnaswamy Stadium             NaN   
..                                         ...             ...   
751                           Wankhede Stadium   Nanda Kishore   
752                  M. A. Chidambaram Stadium     Nigel Llong   
753                           ACA-VDCA Stadium             NaN   
754                           ACA-VDCA Stadium   Sundaram Ravi   
755         Rajiv Gandhi Intl. Cricket Stadium     Nitin Menon   

            umpire2            

In [71]:
df = data.set_index('id')

In [72]:
df.head()

season       city        date                        team1  \
id                                                               
1     2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
2     2017       Pune  2017-04-06               Mumbai Indians   
3     2017     Rajkot  2017-04-07                Gujarat Lions   
4     2017     Indore  2017-04-08       Rising Pune Supergiant   
5     2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                          team2                  toss_winner toss_decision  \
id                                                                           
1   Royal Challengers Bangalore  Royal Challengers Bangalore         field   
2        Rising Pune Supergiant       Rising Pune Supergiant         field   
3         Kolkata Knight Riders        Kolkata Knight Riders         field   
4               Kings XI Punjab              Kings XI Punjab         field   
5              Delhi Daredevils  Royal Challengers Bangalore           bat   

    result  dl_applied                       winner  win_by_runs  \
id                                                                 
1   normal           0          Sunrisers Hyderabad           35   
2   normal           0       Rising Pune Supergiant            0   
3   normal           0        Kolkata Knight Riders            0   
4   normal           0              Kings XI Punjab            0   
5   normal           0  Royal Challengers Bangalore           15   

    win_by_wickets player_of_match                                      venue  \
id                                                                              
1                0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
2                7       SPD Smith    Maharashtra Cricket Association Stadium   
3               10         CA Lynn     Saurashtra Cricket Association Stadium   
4                6      GJ Maxwell                     Holkar Cricket Stadium   
5                0       KM Jadhav                      M Chinnaswamy Stadium   

           umpire1        umpire2 umpire3  
id                                         
1      AY Dandekar       NJ Llong     NaN  
2   A Nand Kishore         S Ravi     NaN  
3      Nitin Menon      CK Nandan     NaN  
4     AK Chaudhary  C Shamshuddin     NaN  
5              NaN            NaN     NaN

In [73]:
df.index

Int64Index([    1,     2,     3,     4,     5,     6,     7,     8,     9,
               10,
            ...
            11342, 11343, 11344, 11345, 11346, 11347, 11412, 11413, 11414,
            11415],
           dtype='int64', name='id', length=756)

In [74]:
df.iloc[1,1]# row with index 1 and all athe cloumns

'Pune'

In [75]:
df.iloc[1,1:3] # row with index 1 and columns with index1,2
# iloc stop index is exclusive

city          Pune
date    2017-04-06
Name: 2, dtype: object

In [76]:
df.loc[1,'season':'city']
# loc -  the stop label also inclusive

season         2017
city      Hyderabad
Name: 1, dtype: object

In [77]:
df.iloc[1:4,1:3]
# rows iwth indices 1,2,3
# cloumns with indices 1,2

city        date
id                    
2     Pune  2017-04-06
3   Rajkot  2017-04-07
4   Indore  2017-04-08

In [78]:
df.loc[1:3,'season':'city']

season       city
id                   
1     2017  Hyderabad
2     2017       Pune
3     2017     Rajkot